lib/ansible/vars/manager.py

```diff
@@ -661,3 +677,51 @@ def set_host_variable(self, host, varname, value):
             self._vars_cache[host] = combine_vars(self._vars_cache[host], {varname: value})
         else:
             self._vars_cache[host][varname] = value
+
+
+class VarsWithSources(MutableMapping):
+    '''
+    Dict-like class for vars that also provides source information for each var
+
+    This class can only store the source for top-level vars. It does no tracking
+    on its own, just shows a debug message with the information that it is provided
+    when a particular var is accessed.
+    '''
+    def __init__(self, *args, **kwargs):
+        ''' Dict-compatible constructor '''
+        self.data = dict(*args, **kwargs)
+        self.sources = {}
+
+    @classmethod
+    def new_vars_with_sources(cls, data, sources):
+        ''' Static class method to instantiate class with sources '''
```

Comment:
Cosmetic:

```suggestion
        ''' Alternate constructor to instantiate class with sources '''
```

A bit more informative about what the method is intended to be used for.